### ISSUE TITLE
Predict method for test set.

### ISSUE THREAD

**Issue by lorenzoFabbri** [williamFalcon/pytorch-lightning#89]
The main Lightning module requires to define the `test_dataloader` function. But I'm not able to find any method that requires the test loader as input. Is there a `model.predict()` method to call on the test set?

**Comment by williamFalcon** [comment]
thanks for bringing that up! we use different scripts for test  set, but makes sense to add a .predict function to trainer.

The proposed flow would be:
```python
# CASE  1: train, then predict
trainer = Trainer(...)
trainer.fit(model)

# the optional prediction call  which  will use  test set
# this makes sure  the  research is 100%  sure  the test set  was  not used in training
# linking  predict to trainer means multi-gpu and cluster  support  for test set  is  free
trainer.test()

# CASE  2: test from saved model
model = CoolModel. load_from_metrics(weights_path='path/to/weights/', tags_csv='path/to/hparams.csv')   
trainer=Trainer()
trainer.test(model)
```

**Comment by lorenzoFabbri** [comment]
Yes, they are. And I like the idea that you can define the `test_step` freely.

Regarding the 3rd case: it's fine, but it might happen that the way you make predictions might be different from the training phase. For instance: you have two images referring to the same object (so same label) and in that case you want to take the *most confident* one among the two.

**Comment by williamFalcon** [comment]
@lorenzoFabbri so define another  method (optional) .predict in cases where your .forward and .predict behavior is different?

so, if no .predict is defined, the model uses .forward. otherwise .predict?

**Comment by lorenzoFabbri** [comment]
Yes, it's a possibility. But I was just wondering whether it was already available. Sometimes you just forget to put `model.eval()` and nothing works as expected: the less code I have to write the better :)

**Comment by williamFalcon** [comment]
not yet, but will add for next release this week. just want to make sure the API and usage are intuitive.

i agree, separate test script is a PITA right now.

**Comment by williamFalcon** [comment]
Otherwise, i'll look at it later this week

**Comment by williamFalcon** [comment]
@lorenzoFabbri  https://gitter.im/PyTorch-Lightning/community?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge

**Comment by fschlatt** [comment]
I've been looking into adding a predict/test function. To have it work like case 2, I think we would need to extract all the setup that is done in `fit` into another function or generalize it somehow

**Comment by expectopatronum** [comment]
Is anyone working on Case 1? I would like to use the functionality soon, so I can give it a try. 
@williamFalcon Could you explain what you mean by 

> and change where the forward calls get routed to (training_step, etc...)

I don't get it :) Or does it simply mean to take care that `validation_end()` is not called when  `test=True`?

**Comment by lorenzoFabbri** [comment]
I'm not working on it at the moment.

**Comment by expectopatronum** [comment]
Thanks, I am working on it!

**Comment by expectopatronum** [comment]
Sure, makes sense!
I guess it would also make sense to rename `validate` to `evaluate` and pass on the flag? (at least from my current understanding of the code)

**Comment by expectopatronum** [comment]
So for testing we can also have multiple test_dataloaders? In the [docu](https://williamfalcon.github.io/pytorch-lightning/LightningModule/RequiredTrainerInterface/#val_dataloader) it's only mentioned for the val_dataloader

**Comment by expectopatronum** [comment]
Great, I agree :)

**Comment by expectopatronum** [comment]
Should I create a WIP pull request or should I wait until I think I am done?

**Comment by expectopatronum** [comment]
For case 2 the current output would look like this:
`1it [00:00, 78.39it/s, batch_nb=0, epoch=0, gpu=0, loss=0.000, test_loss=0.0152, v_nb=0]`
(I am not sure where '1it' comes from yet)
I guess it would make sense to not show `batch_nb`, `epoch`, `loss`? 
Also `v_nb` (what is that by the way?)

**Comment by williamFalcon** [comment]
i think the validation call takes care to not log irrelevant stuff. 1it means 1 iteration. that number changes with the number of batches. 

v_nb is experiment version number. 
there’s a blacklist in the code, let’s just show what doesn’t get blacklisted. 

WIP PR if you want. i’ll review it once the PR is completely done with tests (and has passed tests). then i’ll do a code review at that point

**Comment by expectopatronum** [comment]
I have added 3 test functions:
* `test_multiple_test_dataloader` (analogous to `test_multiple_val_dataloader`)
* `test_multiple_test_dataloader`
* `test_running_test_after_fitting`
Where the last two are only there to check whether the code runs. I couldn't find any code that tests the functionality of the validation functions so I am not sure how to properly test the testing code. Looking forward to suggestions.

**Comment by expectopatronum** [comment]
One thing that could easily be checked is whether it still runs when the model does not have a `test_step` but a `validation_step` by creating a new model.

**Comment by expectopatronum** [comment]
Regarding the documentation: The structure of the Trainer documentation is not clear to me - where should I put the `test()` documentation? Create a new section below `Validation loop` in `index.md`?

**Comment by williamFalcon** [comment]
@expectopatronum awesome PR! very key feature we were missing. Thanks for your work on this!

Merged to master

**Comment by s-rog** [comment]
Is there any documentation on the predict functionalities implemented in the PR?

**Comment by expectopatronum** [comment]
I am not sure what you mean. This PR implements test which is documented [here](https://williamfalcon.github.io/pytorch-lightning/Trainer/Testing%20loop/)

**Comment by s-rog** [comment]
so this test functionality can only return metrics and is not the same as inference/predict?

**Comment by Diyago** [comment]
any updates on a proper infernce with returning predictions?
